lib/modules/manager/nix/extract.ts

```diff
@@ -1,23 +1,138 @@
+import { logger } from '../../../logger';
+import { getSiblingFileName } from '../../../util/fs';
 import { regEx } from '../../../util/regex';
 import { GitRefsDatasource } from '../../datasource/git-refs';
-import { id as nixpkgsVersioning } from '../../versioning/nixpkgs';
 import type { PackageDependency, PackageFileContent } from '../types';
+import { NixFlakeLock } from './schema';
 
-const nixpkgsRegex = regEx(/"github:nixos\/nixpkgs(\/(?<ref>[a-z0-9-.]+))?"/i);
+// TODO: add support to update nixpkgs branches in flakes.nix using nixpkgsVersioning
+
+// as documented upstream
+// https://github.com/NixOS/nix/blob/master/doc/manual/source/protocols/tarball-fetcher.md#gitea-and-forgejo-support
+const lockableHTTPTarballProtocol = regEx(
+  '^https://(?<domain>[^/]+)/(?<owner>[^/]+)/(?<repo>[^/]+)/archive/(?<rev>.+).tar.gz$',
+);
+
+export function extractPackageFile(
+  content: string,
+  packageFile: string,
+): PackageFileContent | null {
+  const packageLockFile = getSiblingFileName(packageFile, 'flake.lock');
+
+  logger.trace(`nix.extractPackageFile(${packageLockFile})`);
 
-export function extractPackageFile(content: string): PackageFileContent | null {
   const deps: PackageDependency[] = [];
 
-  const match = nixpkgsRegex.exec(content);
-  if (match?.groups) {
-    const { ref } = match.groups;
-    deps.push({
-      depName: 'nixpkgs',
-      currentValue: ref,
-      datasource: GitRefsDatasource.id,
-      packageName: 'https://github.com/NixOS/nixpkgs',
-      versioning: nixpkgsVersioning,
-    });
+  const flakeLockParsed = NixFlakeLock.safeParse(content);
+  if (!flakeLockParsed.success) {
+    logger.debug(
+      { packageLockFile, error: flakeLockParsed.error },
+      `invalid flake.lock file`,
+    );
+    return null;
+  }
+
+  const flakeLock = flakeLockParsed.data;
+
+  for (const depName of Object.keys(flakeLock.nodes)) {
+    // the root input is a magic string for the entrypoint and only references other flake inputs
+    if (depName === 'root') {
+      continue;
+    }
+
+    // skip all locked nodes which are not in the flake.nix and cannot be updated
+    // istanbul ignore next: a valid flake.lock file will never run into this
+    if (!(depName in (flakeLock.nodes['root'].inputs ?? []))) {
+      logger.debug(
```

Comment:
so we could do this, not sure if that makes any difference 

```suggestion
    if (!(depName in (flakeLock.nodes['root'].inputs ?? /* istanbul ignore next */ '' []))) {
      logger.debug(
```